packages/riverpod/lib/src/framework/base_provider.dart

```diff
@@ -315,6 +316,43 @@ abstract class ProviderReference {
   /// care about this.
   bool get mounted;
 
+  /// Will return the currently exposed state of the provider.
+  ///
+  /// This is useful when dealing with asynchronous operations.
+  ///
+  /// __Cannot__ be used while building a provider.
+  /// ```dart
+  /// final provider = Provider<int>((ref) {
+  ///   Future.value(1).then((value) => print(ref.currentState));
+  ///   return 0;
+  /// });
+  /// ```
+  Listened get currentState;
+
+  /// Will return the previous exposed state of the provider.
+  ///
+  /// ```dart
+  /// final provider = Provider<int>((ref) {
+  ///   Future.value(1).then((value) => print(ref.previousState));
+  ///   return 0;
+  /// });
+  /// ```
+  Listened? get previousState;
```

Comment:
I'm not sure if it should be added _right now_

There's a bit of thinking involved. The biggest issue is that in many cases, the previous value is technically disposed (StateNotifiers/ChangeNotifiers), so they aren't usable.

It also means that the previous instance cannot be garbage-collected

I'd wait a bit before adding this feature. I'd prefer understanding the use-case more. I mentioned it only because it was in the back of my mind, but it's too early

Reply:
Yes the previous instance lives (even if disposed) as long as the provider does not build again.
What about an explicit opt in? 
```dart
Provider((_) {
/// either T or always null (+ assert because allowPreviousState is false)
ref.previousState
},
  allowPreviousState: true,
);
```

Reply:
That doesn't work with StreamProviders when using `async*`, as the flag would be set asynchronously (even when there's no `await` involved)

It'd need to be something like `Provider.withPrevious(...)`

Reply:
Lets remove `.previousState` from code or should i just delete the public interface to make it unaccessable?

Reply:
Let's remove it